lib/Importer/ONNX.cpp

```diff
@@ -502,6 +502,37 @@ std::unique_ptr<ONNXModelLoader> ONNXModelLoader::parse(const void *onnxModel,
 ONNXModelLoader::ONNXModelLoader(Function &F)
     : CommonOperatorLoader({}, {}, F) {}
 
+void ONNXModelLoader::CheckIO(onnx::GraphProto &net,
+                              llvm::ArrayRef<const char *> names,
+                              llvm::ArrayRef<Tensor *> tensors)
+{
+  for (unsigned i = 0; i < names.size(); i++) {
+    // Look if a corresponding input exists
+    for(int j = 0; j < net.input_size(); j++) {
+      const onnx::ValueInfoProto& value_info = net.input(j);
+      const std::string& input_name = value_info.name();
+
+      if (input_name == names[i]) {
+        llvm::ArrayRef<size_t> dims = tensors[i]->dims();
+        const onnx::TensorShapeProto& shape = value_info.type().tensor_type().shape();
+
+        // Check
+        if (dims.size() != shape.dim_size()) {
+          llvm::outs().flush();
+          llvm_unreachable("Mismatch between input image and ONNX input shape");
+        }
+        // Allow batch dimensions to be different.
```

Comment:
I'm not sure we want to allow batch dimensions to be different... Also, it's not necessarily the case that we're loading a proto with batched input -- the loader is supposed to load any proto of any domain.

Reply:
Models are usually flexible with regard the batch dimension. Currently, when we convert a TensorFlow model into an ONNX model, we get a model specified for a batch of 1. But this model would also work in Glow with a different batch value. This is why I don't think we should check the input batch dimension for now .
- Tensorflow: input shape=(?, 224, 224, 3)
- ONNX:  input shape=(1, 224, 224, 3)

Reply:
What if we're loading a proto that does not have batched input? Or, what if the model is very weird and wants to batch input on the last dimension instead of the first dimension? It might not be an image classification model that's using this class -- e.g. it could be text translation, or anything else. So I don't think always ignoring the first dimension is what we want.

Instead, I'm wondering if perhaps it would make sense to add another argument to the `ONNXModelLoader()` constructor: an `ArrayRef<size_t> allowedDifferentDims`, which is passed into `checkInputs()`, which is used to know when it's OK to skip verifying that a dimension has the same size as in the proto. This way, whatever is creating the `ONNXModelLoader` (e.g. `ImageClassifier`) can pass in dimensions that it knows don't need to match.

What do you think?

Reply:
I see your point: `checkInputs()` is currently in the generic ONNXModelLoader while it's code is actually more specific to `ImageClassifier`. Since the model can have multiple inputs with different numbers of dimensions and since batching is the only use-case I currently see, we can indeed pass an array of `size_t `to express that at most one dimension can be ignore per input tensor during the check.

Reply:
I'm realizing since we don't know the order of the inputs, it probably wouldn't work to just pass the array of size_t -- you can't tell what inputs go with what dimensions. Perhaps it could be an array of pairs with the name of the input and the ignored dimension? So that you can associate different inputs by name with what dimension is OK to be different.